### PR TITLE
Fix libp2p peer connections in the cloud

### PR DESCRIPTION
**Pull request by bitwiseguy**
## Problem description
In the cloud, a node is able to connect to a `bootpeer` because the node is provided the full `multiaddr` (including public ip address) of the `bootpeer`. However, `Anthony` and `Brad` are not able to connect to each other unless one is set as a `bootpeer` of the other.

## Desired behavior
* `Iris` starts with no `bootpeers`
* `Anthony` uses `Iris` as his sole `bootpeer`
* `Brad` uses `Iris` as his sole `bootpeer`
* `Anthony` and `Brad` should be able to connect to each other using the `multiaddrs` stored in the `dht`

## Proposed solution
Theory for why this works in local integration tests but not in the cloud: the nodes all publish their private ip addresses to the `dht` instead of their public ips.

Solution: each node is given its public ip address as part of the node config (or as an env var). The node then uses that public ip address to form its `multiaddr` that gets published to the network, allowing other nodes to determine how to communicate with peers as long as the sender has the recipient's `peerId`.

Some inspiration taken from https://github.com/libp2p/go-libp2p/issues/460

## How was this tested
Created a new Droplet with nitro node called `charley`, who's config was only provided `iris` as a boot peer. Just by `charley` publishing his public IP, he was able to establish peer connections with `anthony` and `brad` as well. Logs from `charley` node:
```
{"time":"2023-09-15T15:25:24.074005349Z","level":"DEBUG","msg":"connected to boot peer","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037","peer":{"ID":"16Uiu2HAmDQj9XnHPD1d4GzS1K3DhmM5Z1cSGLfVTwVcNVdEsAN3Q","Addrs":["/ip4/67.207.88.72/tcp/3005"]}}
{"time":"2023-09-15T15:25:24.07450552Z","level":"INFO","msg":"waiting for bootpeer connections","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037","expectedPeers":1}
{"time":"2023-09-15T15:25:24.175345348Z","level":"DEBUG","msg":"peers found","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037","found-peers":1,"expected-peers":1}
{"time":"2023-09-15T15:25:24.1757988Z","level":"DEBUG","msg":"peer info","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037","peer":"16Uiu2HAmDQj9XnHPD1d4GzS1K3DhmM5Z1cSGLfVTwVcNVdEsAN3Q"}
{"time":"2023-09-15T15:25:24.176156327Z","level":"INFO","msg":"initial threshold for peer connections has been met","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037"}
{"time":"2023-09-15T15:25:24.176448041Z","level":"INFO","msg":"DHT setup complete","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037"}
{"time":"2023-09-15T15:25:24.177189813Z","level":"INFO","msg":"Initializing chain service..."}
{"time":"2023-09-15T15:25:24.277057561Z","level":"DEBUG","msg":"Adding state channel address to dht","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037"}
{"time":"2023-09-15T15:25:24.281798201Z","level":"INFO","msg":"Added state channel address to dht","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037"}
{"time":"2023-09-15T15:25:24.760833326Z","level":"INFO","msg":"Connected to ethclient","url":"wss://wss.calibration.node.glif.io/apigw/lotus/rpc/v0"}
{"time":"2023-09-15T15:25:24.921969215Z","level":"INFO","msg":"Found chain id","chainId":314159}
{"time":"2023-09-15T15:25:25.407697909Z","level":"INFO","msg":"checking for missed chain events","address":"0x6cf836B2A139e04FF05A0E8faA47b5731E331d48","startBlock":915502,"currentBlock":915503}
{"time":"2023-09-15T15:25:25.659768101Z","level":"INFO","msg":"finished checking for missed chain events","address":"0x6cf836B2A139e04FF05A0E8faA47b5731E331d48","numMissedEvents":0}
{"time":"2023-09-15T15:25:25.820627608Z","level":"INFO","msg":"Constructed Engine","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037"}
{"time":"2023-09-15T15:25:25.821068168Z","level":"INFO","msg":"Initializing Http RPC transport..."}
{"time":"2023-09-15T15:25:25.821589796Z","level":"INFO","msg":"Completed RPC server initialization"}
Hosting UI at http://localhost:5005/
{"time":"2023-09-15T15:26:13.263308184Z","level":"DEBUG","msg":"notification: connected to peer","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037","peerId":"16Uiu2HAmMZD9hwtMBBDpbE3qEigb4FF1YDAjePLCVR6RQZYo6RPJ","peerCount":2}
{"time":"2023-09-15T15:26:17.877251185Z","level":"DEBUG","msg":"notification: connected to peer","address":"0xbb01cf80eFe3CDFdFC0c047e39D1aF8a7Ed14037","peerId":"16Uiu2HAmUPifMQSzPGmt3UdTc86bMuebBJ26e2xBWus7xmQ5zhMt","peerCount":3}
```

### CODE REVIEW

File: node/engine/messageservice/p2p-message-service/service.go [hunk 1]
```diff
@@ -103,25 +111,34 @@ func NewMessageService(ip string, port int, me types.Address, pk []byte, bootPee
...
+
+		peer, err := peer.AddrInfoFromP2pAddr(addr)
+		ms.checkError(err)
+
+		bootAddrs = append(bootAddrs, *peer)
+	}
+
 	var options []dht.Option
 	options = append(options, dht.BucketSize(20))
+	options = append(options, dht.BootstrapPeers(bootAddrs...))
```
Comment (bitwiseguy) [inline node/engine/messageservice/p2p-message-service/service.go:141]: From GoDoc: `BootstrapPeers` configures the bootstrapping nodes that we will connect to to seed and refresh our Routing Table if it becomes empty.

**Comment by geoknee** [comment]
Once this is merged to `main` we need to consider getting the changes into our demo feature branch as well (and undoing any short term fixes that we had in place).

File: .github/workflows/deploy-cloud-nodes.yml [hunk 2]
```diff
@@ -64,7 +64,14 @@ jobs:
             docker pull registry.digitalocean.com/magmo/go-nitro:latest
             docker stop $NODE_NAME || true
             docker rm $NODE_NAME || true
-            docker run --restart=unless-stopped -it -d --name $NODE_NAME -p 3005:3005 -p 4005:4005 -p 80:5005 -e NITRO_CONFIG_PATH=$NITRO_CONFIG_PATH -e SC_PK=$SC_PK -e CHAIN_PK=$CHAIN_PK -v /var/nitro_store:/app/data registry.digitalocean.com/magmo/go-nitro:latest
+            docker run --restart=unless-stopped -it -d --name $NODE_NAME \
+              -p 3005:3005 -p 4005:4005 -p 80:5005 \
+              -e NITRO_PUBLIC_IP="0.0.0.0"
```
Comment (kerzhner) [inline .github/workflows/deploy-cloud-nodes.yml]: Curious, why is the public address set to "0.0.0.0"?

File: .github/workflows/deploy-cloud-nodes.yml [hunk 3]
```diff
@@ -64,7 +64,14 @@ jobs:
             docker pull registry.digitalocean.com/magmo/go-nitro:latest
             docker stop $NODE_NAME || true
             docker rm $NODE_NAME || true
-            docker run --restart=unless-stopped -it -d --name $NODE_NAME -p 3005:3005 -p 4005:4005 -p 80:5005 -e NITRO_CONFIG_PATH=$NITRO_CONFIG_PATH -e SC_PK=$SC_PK -e CHAIN_PK=$CHAIN_PK -v /var/nitro_store:/app/data registry.digitalocean.com/magmo/go-nitro:latest
+            docker run --restart=unless-stopped -it -d --name $NODE_NAME \
```
Comment (kerzhner) [inline .github/workflows/deploy-cloud-nodes.yml:67]: Perhaps to be addressed in another PR, but there is a lot of copy/pasting in this file.

**Review by kerzhner — APPROVED** [review]

File: .github/workflows/deploy-cloud-nodes.yml (hunk 2 above)
Comment (kerzhner) [reply .github/workflows/deploy-cloud-nodes.yml]: Looks like this might have been addressed in https://github.com/statechannels/go-nitro/pull/1711/commits/5fc35f02bfc3c36aa6684a30c0f96624083921d2?

Comment (bitwiseguy) [reply .github/workflows/deploy-cloud-nodes.yml]: That was an artifact of some troubleshooting I was doing. That line has been changed to the following:
```
              -e NITRO_PUBLIC_IP=$DROPLET_IP
```

File: .github/workflows/deploy-cloud-nodes.yml (hunk 3 above)
Comment (bitwiseguy) [reply .github/workflows/deploy-cloud-nodes.yml:67]: Yes I agree. Ideally this would be wrapped in a function within the `.yml` github workflow file. I'm not sure if that's possible. Could also look into copying a `makefile` into each of the droplets and wrapping all these `docker` commands into a single `make` command

Comment (bitwiseguy) [reply .github/workflows/deploy-cloud-nodes.yml:67]: I propose we make that optimization in a future PR

Comment (kerzhner) [reply .github/workflows/deploy-cloud-nodes.yml:67]: Would composite actions help here? https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

Comment (bitwiseguy) [reply .github/workflows/deploy-cloud-nodes.yml:67]: I'm not sure that would solve it but perhaps I'm not understanding the full capability of composite actions. The tricky part of our workflow is that we need those docker commands to run within the context of the ssh session inside the Droplet
